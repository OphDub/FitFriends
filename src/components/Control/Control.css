.Control {
  display: flex;
  flex-direction: column;
  width: 300px; }

.Control-input {
  margin-bottom: 15px; }
