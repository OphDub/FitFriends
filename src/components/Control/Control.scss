.Control {
  display: flex;
  flex-direction: column;
  width: 300px;
}

.Control-input {
  margin-bottom: 15px;
}

.login-btn {
  border: none;
}

.login-btn:hover {
  background-color: #48E0EA;
  color: white;
}