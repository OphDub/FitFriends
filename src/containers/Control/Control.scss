.Control {
  display: flex;
  flex-direction: column;
  width: 300px;
}

.Control-input {
  margin-bottom: 15px;
  padding: 5px;
  border-radius: 5px;
  border: none;
  outline: none;
}

.login-btn {
  border: none;
  padding: 8px;
  text-transform: uppercase;
  border-radius: 5px;
  outline: none;
}

.login-btn h3 {
  margin: 0px;
}

.login-btn:hover {
  background-color: #48E0EA;
  color: white;
  cursor: pointer;
}