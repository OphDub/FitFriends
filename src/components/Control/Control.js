import React, { Component } from 'react';
import './Control.css';
import { loginUser } from '../../actions/actionsIndex';
import { connect } from 'react-redux';

export class Control extends Component {
  constructor () {
    super ();
    this.state = {
      username: '',
      password: '',
    }
  }

  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  }

  handleLogin = (e) => {
    e.preventDefault();

    //authenticate user here
    this.props.loginUser(this.state);
    this.props.handleReroute();
    this.setState({ username: '', password: '' });
  }

  render () {
    return (
      <form type="submit" className="Control">
        <input  className="Control-input Control-username"
          type="email"
          placeholder="Email"
          name="username"
          value={this.state.username}
          onChange={this.handleChange}/>
        <input  className="Control-input Control-password"
          type="password"
          placeholder="Password"
          name="password"
          value={this.state.password}
          onChange={this.handleChange}/>
        <button className="login-btn"
          onClick={this.handleLogin}
          type="submit">
            Login
        </button>
      </form>
    )
  }
};

export const mapStateToProps = (state) => ({
  history: state.history
})

export const mapDispatchToProps = (dispatch) => ({
  loginUser: (user) => dispatch(loginUser(user))
})

export default connect(mapStateToProps, mapDispatchToProps)(Control);