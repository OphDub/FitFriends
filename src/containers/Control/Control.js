import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import './Control.css';
import { login } from '../../thunks/thunks';
import { withRouter } from 'react-router-dom';

export class Control extends Component {
  constructor () {
    super();
    this.state = {
      email: '',
      password: '',
      errorMsg: ''
    };
  }

  handleChange = (e) => {
    const { name, value } = e.target;
    this.setState({ [name]: value });
  }

  validateLogin = () => {
    const { email, password } = this.state;

    const message = 'Please provide an email and password.';
    let errorMsg;

    const validated = email === '' || password === '' ?
      (errorMsg = message) && false : true;

    this.setState({ errorMsg });
    return validated;
  }

  handleLogin = async (e) => {
    e.preventDefault();
    const { email, password } = this.state;

    if (!this.validateLogin()) {
      return;
    }

    try {
      await this.props.login(email, password);

      this.setState({
        email: '',
        password: '',
        errorMsg: ''
      });

      this.props.history.push('/workout');
    } catch (error) {
      const errorMsg = 'Username/password incorrect. Try again.';

      this.setState({
        email: '',
        password: '',
        errorMsg
      });

      throw error;
    }
  }

  renderError = () => {
    return (<h3>{this.state.errorMsg}</h3>);
  }

  render () {
    return (
      <form type="submit" className="Control">
        <input  className="Control-input Control-username"
          type="email"
          placeholder="Email"
          name="email"
          value={this.state.email}
          onChange={this.handleChange}/>
        <input  className="Control-input Control-password"
          type="password"
          placeholder="Password"
          name="password"
          value={this.state.password}
          onChange={this.handleChange}/>
        <button className="login-btn"
          onClick={this.handleLogin}
          type="submit">
          <h3>Login</h3>
        </button>
        <div className="error-msg">
          {this.renderError()}
        </div>
      </form>
    );
  }
}

const historyProps = PropTypes.shape({
  push: PropTypes.func
});

Control.propTypes = {
  login: PropTypes.func.isRequired,
  history: historyProps
};

export const mapDispatchToProps = (dispatch) => ({
  login: (email, password) => dispatch(login(email, password))
});

export default withRouter(connect(null, mapDispatchToProps)(Control));