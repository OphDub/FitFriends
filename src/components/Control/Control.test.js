import React from 'react';
import { shallow } from 'enzyme';
import { Control, mapStateToProps, mapDispatchToProps } from './Control';

describe('CONTROL', () => {
  let renderedComponent;

  beforeEach(() => {
    renderedComponent = shallow(<Control />);
  });

  it('should match snapshot', () => {
    expect(renderedComponent).toMatchSnapshot();
  });

  it('should change state when handleChange is called', () => {
    const mockEvent = { target: { value: 'hello', name: 'username'}};
    const expected = 'hello';

    renderedComponent.instance().handleChange(mockEvent);
    renderedComponent.update();

    expect(renderedComponent.state().username).toEqual(expected);
  });

  it('should clear state when handleLogin is called', () => {
    const exampleString = 'hello';
    const exampleEvent = { target: { value: 'hello', name: 'username'}};
    const mockFn = jest.fn();
    const wrapper = shallow(<Control loginUser={mockFn} handleReroute={mockFn}/>);

    wrapper.instance().handleChange(exampleEvent);
    wrapper.update();

    expect(wrapper.state().username).toEqual(exampleString);

    const expected = '';
    const mockEvent = {preventDefault: jest.fn()};

    wrapper.instance().handleLogin(mockEvent);
    wrapper.update();

    expect(wrapper.state().username).toEqual(expected);
  });

  it('calls loginUser and handleReroute when handleLogin is called', () => {
    const mockFn = jest.fn();
    const wrapper = shallow(<Control loginUser={mockFn} handleReroute={mockFn}/>);
    const mockEvent = { preventDefault: jest.fn() };

    wrapper.instance().handleLogin(mockEvent);
    expect(mockFn).toHaveBeenCalled();
  });

  it('should call the dispatch func when using a func from mapDispatchToProps', () => {
    const mockDispatch = jest.fn();
    const mapped = mapDispatchToProps(mockDispatch);

    mapped.loginUser();

    expect(mockDispatch).toHaveBeenCalled();
  })
});
