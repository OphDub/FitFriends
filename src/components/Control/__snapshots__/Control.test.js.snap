// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CONTROL should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Control />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          className="Control-input Control-username"
          name="username"
          onChange={[Function]}
          placeholder="username"
          type="text"
          value=""
/>,
        <input
          className="Control-input Control-password"
          name="password"
          onChange={[Function]}
          placeholder="password"
          type="password"
          value=""
/>,
        <button
          className="login-btn"
          onClick={[Function]}
>
          Login
</button>,
      ],
      "className": "Control",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "Control-input Control-username",
          "name": "username",
          "onChange": [Function],
          "placeholder": "username",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "className": "Control-input Control-password",
          "name": "password",
          "onChange": [Function],
          "placeholder": "password",
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Login",
          "className": "login-btn",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Login",
        "type": "button",
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            className="Control-input Control-username"
            name="username"
            onChange={[Function]}
            placeholder="username"
            type="text"
            value=""
/>,
          <input
            className="Control-input Control-password"
            name="password"
            onChange={[Function]}
            placeholder="password"
            type="password"
            value=""
/>,
          <button
            className="login-btn"
            onClick={[Function]}
>
            Login
</button>,
        ],
        "className": "Control",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "Control-input Control-username",
            "name": "username",
            "onChange": [Function],
            "placeholder": "username",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "className": "Control-input Control-password",
            "name": "password",
            "onChange": [Function],
            "placeholder": "password",
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Login",
            "className": "login-btn",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Login",
          "type": "button",
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
